### ISSUE TITLE
push replication sometimes doesn't work if user backgrounds the app quickly

### ISSUE THREAD

**Issue by npomfret** [couchbase/couchbase-lite-ios#1475]
It seems that if a user pressed the home button immediately after creating/updating a document then the continuous push replication process sometimes doesn't fire quickly enough to push the data to the server.

I understand that there's a small delay. (IMO) Given that the developer has the option to bulk insert I would suggest that the delay is not required.

**Comment by snej** [comment]
Seems like when the pusher is in this state (revs in its inbox queue but not yet sent in a revs_diff request) it's not considered 'active' enough to trigger creating a background session when the app is backgrounded. I'll look at the state logic.

Originally I added this queueing delay because the notifications from a bulk insert on the main thread might not arrive all at once on the replicator's bg thread. But I'm pretty certain that's no longer the case — a thread suppresses notifications while a transaction is open and then sends out a single notification with all the changes. So it should be safe to process them immediately when received.

**Comment by snej** [comment]
(Added cross-platform tag because the pusher's inbox queue might still be in the Java or C# ports too.)

**Comment by snej** [comment]
From further discussion in the [forum thread](https://forums.couchbase.com/t/is-there-a-way-to-perform-instant-push-replication/10201/8), it sounds like the process was suspended before the HTTP response was delivered back to the app code. (Or perhaps after that but before the app could send a request to start replication?)

Turning on `Router` logging will help narrow this down, since it will log to the console when an HTTP request is received and the response is sent.

The HTTP router may need to listen for backgrounding notifications and create its own background sessions, to provide time for a response to be sent. (Unless the request is one that can wait indefinitely or never returns, like a changes feed.)

**Comment by npomfret** [comment]
I noticed the needs more info.  Can I help??  To give some more context...

What I've noticed is that there's a delay between a user making a document change and a continuous-replication task pushing that change to the server.  The delay is large enough that if the user presses the home button right after making the change then it doesn't reach the server until the app is brought back to the foreground.

I've tried explicitly starting a push replication task after each db write, but that didn't work.  And then I tried starting a push replication task when the app detects that it is going into the background.  Strangely that didn't work either.  The task doesn't appear to start at all until the app is back in the foreground.

I'm using the REST api in a react-native application.

Is there any other information I can provide?
